### PR TITLE
Add SecurityContext to delegating TaskScheduler

### PR DESCRIPTION
**Pull request by gbaso**
Wrap DelegatingSecurityContextTaskScheduler's Runnable tasks in
DelegatingSecurityContextRunnables, allowing to specify a
SecurityContext to use for tasks execution.

Closes gh-9514

### CODE REVIEW

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java [hunk 1]
```diff
@@ -30,47 +35,64 @@
  * @author Richard Valdivieso
  * @since 5.1
  */
-public class DelegatingSecurityContextTaskScheduler implements TaskScheduler {
+public class DelegatingSecurityContextTaskScheduler extends AbstractDelegatingSecurityContextSupport
```
Comment (jzheaux) [inline core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java:38]: Ideally, this class should favor composition over inheritance. Since `DelegatingSecurityContextCallable` and `DelegatingSecurityContextRunnable` have factory methods that do what you need, would you be able to inline the support's `wrap` methods here in `DelegatingSecurityContextTaskScheduler`?

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java [hunk 2]
```diff
@@ -30,47 +35,64 @@
...
+		implements TaskScheduler {
 
-	private final TaskScheduler taskScheduler;
+	private final TaskScheduler delegate;
```
Comment (jzheaux) [inline core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java:41]: Since this variable name change is not necessary to close the ticket, would you please move this change to a separate commit in this PR?

File: core/src/test/java/org/springframework/security/scheduling/CurrentDelegatingSecurityContextTaskSchedulerTests.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Tests using the current {@link SecurityContext} on
+ * {@link DelegatingSecurityContextTaskScheduler}
+ *
+ * @author Giacomo Baso
+ * @since 5.5
+ *
+ */
+public class CurrentDelegatingSecurityContextTaskSchedulerTests
```
Comment (jzheaux) [inline core/src/test/java/org/springframework/security/scheduling/CurrentDelegatingSecurityContextTaskSchedulerTests.java:31]: Is there a reason that these task scheduler test classes should be separated like this? Most Spring Security test classes follow the pattern `XXXTests` where `XXX` is the class under test.

File: core/src/test/java/org/springframework/security/scheduling/CurrentDelegatingSecurityContextTaskSchedulerTests.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+		extends AbstractDelegatingSecurityContextTaskSchedulerTests {
+
+	@Before
+	public void setUp() throws Exception {
+		super.currentSecurityContextPowermockSetup();
```
Comment (jzheaux) [inline core/src/test/java/org/springframework/security/scheduling/CurrentDelegatingSecurityContextTaskSchedulerTests.java:36]: While I realize it was done this way in the past, I don't think that mocking the `create` methods for `DelegatingSecurityContextCallable` and `DelegatingSecurityContextRunnable` is necessary.

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks for the PR, @gbaso! I've left some feedback inilne.

File: core/src/test/java/org/springframework/security/scheduling/CurrentDelegatingSecurityContextTaskSchedulerTests.java (hunk 3 above)
Comment (gbaso) [reply core/src/test/java/org/springframework/security/scheduling/CurrentDelegatingSecurityContextTaskSchedulerTests.java:31]: Other related test classes follow the same pattern, in particular `AbstractDelegatingSecurityContextExecutorTests` and subclasses. This allows repeating the tests with a different configuration, i.e. to use the current `SecurityContext` or a provided one.

Would you suggest a different approach?

Comment (gbaso) [reply core/src/test/java/org/springframework/security/scheduling/CurrentDelegatingSecurityContextTaskSchedulerTests.java:31]: Given that those test patterns do not actually check for the use of a specific `SecurityContext` and your comment regarding the mocking of `DelegatingSecurityContextCallable#create`, I've decided to revert to the old tests suite `DelegatingSecurityContextTaskSchedulerTests`, with the addition of a mocked `SecurityContext`.

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java (hunk 2 above)
Comment (gbaso) [reply core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java:39]: Moved to ba53e5d

File: core/src/test/java/org/springframework/security/scheduling/CurrentDelegatingSecurityContextTaskSchedulerTests.java (hunk 4 above)
Comment (gbaso) [reply core/src/test/java/org/springframework/security/scheduling/CurrentDelegatingSecurityContextTaskSchedulerTests.java]: See https://github.com/spring-projects/spring-security/pull/9532#discussion_r608824828

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java (hunk 1 above)
Comment (gbaso) [reply core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java]: wrap is not inlined in 1851a42

File: core/src/test/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskSchedulerTests.java [hunk 5]
```diff
@@ -58,7 +61,8 @@
 	@Before
 	public void setup() {
 		MockitoAnnotations.initMocks(this);
-		this.delegatingSecurityContextTaskScheduler = new DelegatingSecurityContextTaskScheduler(this.scheduler);
+		this.delegatingSecurityContextTaskScheduler = new DelegatingSecurityContextTaskScheduler(this.scheduler,
```
Comment (jzheaux) [inline core/src/test/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskSchedulerTests.java:64]: Should there be a test that confirms that the security context was indeed used?

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java [hunk 6]
```diff
@@ -32,45 +36,66 @@
...
+		return this.delegate.scheduleWithFixedDelay(wrap(task), startTime, delay);
 	}
 
 	@Override
 	public ScheduledFuture<?> scheduleWithFixedDelay(Runnable task, long delay) {
-		return this.taskScheduler.scheduleWithFixedDelay(task, delay);
+		return this.delegate.scheduleWithFixedDelay(wrap(task), delay);
+	}
+
+	protected final Runnable wrap(Runnable delegate) {
```
Comment (jzheaux) [inline core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java:97]: Let's please leave this private for now until there is a clear use case for making it protected. Users who extend this class can easily create their own wrap method.

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks for the updates, @gbaso! I've left some additional feedback inline.

File: core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java (hunk 6 above)
Comment (gbaso) [reply core/src/main/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskScheduler.java]: `wrap` is now private

File: core/src/test/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskSchedulerTests.java (hunk 5 above)
Comment (gbaso) [reply core/src/test/java/org/springframework/security/scheduling/DelegatingSecurityContextTaskSchedulerTests.java:72]: I've added two tests in d455211, for explicit/current security context

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks for the updates, @gbaso!

Thanks for your descriptive commit messages. To simplify maintenance, will you please squash your commits into a single atomic commit?

**Comment by gbaso** [comment]
@jzheaux PR should be ready now

**Comment by jzheaux** [comment]
Thanks, @gbaso! I've scheduled this for 5.6.0-M1.

**Comment by jzheaux** [comment]
Thanks again, @gbaso! This is now merged into `main` via 80743a267c91a8556e3e2c59e698d9b8ee4fc199. I also added a small polish to add the `@since` attribute to the new constructor's JavaDoc via e1e31939a34551a74e5d89318dcc240e7218f176
